azure-iot-hub-devicesdk/tests/test_sync_clients.py

```diff
@@ -0,0 +1,130 @@
+import pytest
+from azure.iot.hub.devicesdk import DeviceClient, ModuleClient
+from azure.iot.hub.devicesdk.transport.mqtt import MQTTTransport
+from azure.iot.hub.devicesdk.transport.abstract_transport import AbstractTransport
+from azure.iot.hub.devicesdk import Message
+
+# Note that the auth_provider fixture is implicitly imported from tests/conftest.py
+
+
+class FakeTransport(AbstractTransport):
+    def __init__(self):
+        pass
+
+    def connect(self, callback):
+        callback()
+
+    def send_event(self, event, callback):
+        callback()
+
+    def send_output_event(self, event, callback):
+        callback()
+
+    def disconnect(self, callback):
+        callback()
+
+
+@pytest.fixture
+def transport(mocker):
+    return mocker.MagicMock(wraps=FakeTransport())
+
+
+class TestClientSharedApi(object):
+    xfail_notimplemented = pytest.mark.xfail(raises=NotImplementedError, reason="Unimplemented")
+
+    client_classes = [DeviceClient, ModuleClient]
+
+    @pytest.fixture(params=client_classes)
+    def client(self, request, transport):
+        client_class = request.param
+        return client_class(transport)
+
+    @pytest.mark.parametrize(
```

Comment:
>@pytest.mark.parametrize( [](start = 4, length = 25)

The verbose output looks good and we should change the build step for pytest to have verbose in my opinion, otherwise its pretty difficult to understand if all cases were covered or not.